Make ToCss' SequenceWriter not monomorphise like a maniac anymore

nox opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19838)
<!-- Reviewable:end -->

nox committed:
Use Image::to_css in <Image as Debug>::fmt

nox committed:
Derive ToCss for font_feature_values_rule::SingleValue

nox committed:
Change ToCss to take a CssWriter<W>

This more concrete wrapper type can write a prefix the very first time something
is written to it. This allows removing plenty of useless monomorphisations caused
by the former W/SequenceWriter<W> pair of types.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/values/computed/position.rs, ports/geckolib/glue.rs, components/style/stylesheets/style_rule.rs, components/style/properties/shorthand/background.mako.rs, components/style/gecko/url.rs and 83 more
 * @canaltinova: components/style/values/computed/position.rs, components/style/stylesheets/style_rule.rs, components/style/properties/shorthand/background.mako.rs, components/style/gecko/url.rs, components/style/values/computed/basic_shape.rs and 82 more
 * @emilio: components/style/values/computed/position.rs, ports/geckolib/glue.rs, components/style/stylesheets/style_rule.rs, components/style/properties/shorthand/background.mako.rs, components/style/gecko/url.rs and 83 more
 * @fitzgen: components/script/dom/cssstyledeclaration.rs, components/script/dom/mediaquerylist.rs, components/script/dom/medialist.rs
 * @kichjang: components/script/dom/cssstyledeclaration.rs, components/script/dom/mediaquerylist.rs, components/script/dom/medialist.rs
 * @asajeffrey: components/script/dom/cssstyledeclaration.rs, components/script/dom/mediaquerylist.rs, components/script/dom/medialist.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style and script code, but no tests are modified. Please consider adding a test!

nox commented:
Cc @emilio @SimonSapin @bholley

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a0cf705fdfd32be1ba62a21ec6dc3e2c5f06291a with merge 2e32dbf662894e2a9b00e7279589f1e40f9409e1...

emilio commented on ports/geckolib/glue.rs:
```diff
@@ -707,8 +707,12 @@ pub extern "C" fn Servo_AnimationValue_Serialize(value: RawServoAnimationValueBo
-                             None, None /* No extra custom properties */);
+        .single_value_to_css(
+            &get_property_id_from_nscsspropertyid!(property, ()),
+            &mut CssWriter::new(&mut string),
```
We should make this pass `buffer` directly. Mind filing an issue? Though I don't recall if @bholley's patch fixed this.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7643)

emilio commented on components/style_traits/values.rs:
```diff
@@ -125,89 +172,15 @@ where
-            has_written: false,
+        let has_prefix = self.inner.prefix.is_some();
+        if !has_prefix {
+            self.inner.prefix = Some(self.separator);
```
oh, smart :)

emilio commented on components/style_traits/values.rs:
```diff
@@ -72,11 +72,58 @@ where
+where
+    W: Write,
+{
+    /// Creates a new `CssWriter`.
```
nit: Maybe `#[inline]`

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4743dc32a198d2bbca234d5cea92ab572bebe77d with merge 3ff184c6f80620b45084250c5dd63f2e80ec64c0...

emilio commented on components/style_traits/values.rs:
```diff
@@ -72,11 +72,58 @@ where
+/// A writer tailored for serialising CSS.
+///
+/// Coupled with SequenceWriter, this allows callers to transparently handle
+/// things like comma-separated values etc.
```
Maybe put an example of how this is used? I had to read it a bit carefully.

In particular, what state does `prefix` represent? If I'm not wrong, when writing a sequence:

 * We start with prefix = `Some("")`.
 * On the first `item`, assuming it writes something, it `take`s the prefix.
 * On the second one, since there's no prefix, we set `prefix: Some(sequence_separator)`, and reset it after the `to_css` call begins.
 * ...

emilio commented on components/style_traits/values.rs:
```diff
@@ -125,89 +172,15 @@ where
-            self.has_written = true;
+        item.to_css(&mut self.inner)?;
+        if !has_prefix {
+            self.inner.prefix = None;
```
I think it'd be nice, more as a self-documenting assert than as a sanity-check, that in this case `self.inner.prefix` is either `None` or `self.separator` before the assignment (since even though it can change down the line, it should always be mutated by another `SequenceWriter`, and thus resetted too.

emilio reviewed (approved):
/me likes :)

Not sure if you want to wait for @SimonSapin's / @bholley's review or not, if you don't just feel free to land with r=me.

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/10763)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0bf70dcc40aa6c176701e6a198b7716038bf1b9d with merge f93f875afe86d6456046c911254ea7f5d304cdca...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10474), [arm32](http://build.servo.org/builders/arm32/builds/10712), [arm64](http://build.servo.org/builders/arm64/builds/10737), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10765), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7630), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7646), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10844), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3246), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3267), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7513), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7578), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3338), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3286), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6147)
State: approved= try=True

nox committed:
Derive ToCss for TransformOperation

Now that SequenceWriter<W> does not monomorphise excessively, we can actually
type check a derived ToCss without too much type recursion.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19841) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 04e7cce1658e78553f3cec98403993c55a7a5cf3 with merge 30957343750c612dbb3b2154e95845c9d7bc0f2b...

nox commented:
https://treeherder.mozilla.org/#/jobs?repo=try&revision=6d184c6e82cdec7dddbbad705896b4fe0b8dbc95

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10857)

nox commented:
https://treeherder.mozilla.org/#/jobs?repo=try&revision=de052cfb29450e5f4a4d5a792c193a0757f2f417

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 42c8dc983f86371e9e89818a2e8d9ad8b24a9900 with merge e9b8fead68cf681e46b4067becc90233ee33a3c2...

bors-servo commented:
:broken_heart: Test failed - [arm32](http://build.servo.org/builders/arm32/builds/10726)

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 42c8dc9 has been approved by `emilio`

<!-- @bors-servo r=emilio 42c8dc983f86371e9e89818a2e8d9ad8b24a9900 -->

nox commented:
@bors-servo p=7

bors-servo commented:
:hourglass: Testing commit 42c8dc983f86371e9e89818a2e8d9ad8b24a9900 with merge 6b2e5283c9e810ac316c203849b6ea25544a62c7...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10493), [arm32](http://build.servo.org/builders/arm32/builds/10735), [arm64](http://build.servo.org/builders/arm64/builds/10757), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10788), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7654), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7670), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10866), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3266), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3287), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7532), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7600), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3360), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3307), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6170)
Approved by: emilio
Pushing 6b2e5283c9e810ac316c203849b6ea25544a62c7 to master...

ionutgoldan commented:
Good news! This noticeably reduced build times on Windows.

== Change summary for alert #11256 (as of Tue, 23 Jan 2018 16:57:02 GMT) ==

Improvements:

  9%  build times windows2012-64 opt rusttests taskcluster-c4.4xlarge     2,321.94 -> 2,103.34
  9%  build times windows2012-32 opt rusttests taskcluster-c4.4xlarge     2,302.99 -> 2,090.30
  9%  build times windows2012-32 opt taskcluster-c4.4xlarge               2,580.50 -> 2,348.30
  9%  build times windows2012-64 opt taskcluster-c4.4xlarge               2,664.19 -> 2,437.35
  8%  build times windows2012-64 opt static-analysis taskcluster-c4.4xlarge2,580.59 -> 2,369.24
  8%  build times windows2012-32 opt static-analysis taskcluster-c4.4xlarge2,255.28 -> 2,079.92
  4%  build times windows2012-32 pgo taskcluster-c4.4xlarge               4,346.07 -> 4,185.03

For up to date results, see: https://treeherder.mozilla.org/perf.html#/alerts?id=11256
